### ISSUE TITLE
[Bug] Specifying multiple decryption certificates

### ISSUE THREAD

**Issue by husseinkorly** [AzureAD/microsoft-identity-web#1243]
**Which version of Microsoft Identity Web are you using?**
1.12.0

**Where is the issue?**
* Web app
    * [ ] Sign-in users
    * [ ] Sign-in users and call web APIs
* Web API
    * [ x] Protected web APIs (validating tokens)
    * [ ] Protected web APIs (validating scopes)
    * [ ] Protected web APIs call downstream web APIs
* Token cache serialization
     * [ x] In-memory caches
     * [ ] Session caches
     * [ ] Distributed caches
* Other (please describe)

**Is this a new or an existing app?**
The app is in production and I have upgraded to a new version of Microsoft Identity Web.

**Repro**
Configuring multiple certificates for decryption and only one of them is valid and it can be used. 

```json
"TokenDecryptionCertificates": [
      {
        "SourceType": "StoreWithThumbprint",
        "CertificateStorePath": "LocalMachine/My",
        "CertificateThumbprint": "962D129A...D18EFEB6961684"
      },
      {
        "SourceType": "StoreWithThumbprint",
        "CertificateStorePath": "LocalMachine/My",
        "CertificateThumbprint": "962D129A...D18EFEB6960000"
      }
    ]
```

**Expected behavior**
A valid certificate (from list of certificates) will be used to decrypt incoming tokens

**Actual behavior**
Fails to decrypt a token if any of the provided certificate is invalid 

**Possible solution**
Try to decrypt the token with any valid certificate provided.

**Additional context / logs / screenshots**
n/a

**Comment by jmprieur** [comment]
@husseinkorly are you saying this is a regression?

**Comment by jmprieur** [comment]
I think I understand what happens.
Now, when the certificate has expired we set it to null in the CertificateDescription (this was done when implementing the certificate rotation). But Microsoft.IdentityModel does not accept a null certificate.

We should only do the yield:

https://github.com/AzureAD/microsoft-identity-web/blob/37471e7578b334bb925c9170e456b05de83368e1/src/Microsoft.Identity.Web/CertificateManagement/DefaultCertificateLoader.cs#L307

when `certDescription.Certificate` is not null

**Comment by jmprieur** [comment]
@jennyf19 is working on a fix. 
Meanwhile, I think that version 1.9.2 should work fine in your case.

**Comment by husseinkorly** [comment]
Thank you. I tried using version 1.9.2 but the API still failing to decrypt the token.

**Comment by jmprieur** [comment]
Can you tell us more about the exception? https://github.com/AzureAD/microsoft-identity-web/wiki/Logging

**Comment by husseinkorly** [comment]
I pulled the new changes in master and I still see the same issue. I am using 2 decryption certificates. One of them is a valid certificate (used to encrypt) and another one that still valid but not used to encrypt the token. I see the API still returning 401 and I don't see any exceptions being logged. I think it's related to this issue with identityModel package that doesn't accept a list of decryption keys
https://github.com/AzureAD/azure-activedirectory-identitymodel-extensions-for-dotnet/issues/1431

**Comment by jennyf19** [comment]
@husseinkorly they both have to be decryption certs, not encryption certs. See this [page](https://github.com/AzureAD/microsoft-identity-web/wiki/Token-Decryption#in-the-azure-ad-portal) -> `"usage": "Encrypt"`

if you're using a client credential certificate, like instead of a secret, a cert, it can go in the [configuration](https://github.com/AzureAD/microsoft-identity-web/wiki/Certificates#describing-client-certificates-to-use-by-configuration).

**Comment by husseinkorly** [comment]
Yes. I am only interested in the decryption cert I have both in the configuration like this for example:

```json
"TokenDecryptionCertificates": [
      {
        "SourceType": "StoreWithThumbprint",
        "CertificateStorePath": "LocalMachine/My",
        "CertificateThumbprint": "962D129A...D18EFEB6961684"
      },
      {
        "SourceType": "StoreWithThumbprint",
        "CertificateStorePath": "LocalMachine/My",
        "CertificateThumbprint": "962D129A...D18EFEB6960000"
      }
    ]
```

One of them is a valid decryption cert (where I have the .pfx installed in my local machine) and it can decrypt the token just fine, but when I add another cert to the list that's just any other cert the API start returning 401.

**Comment by jennyf19** [comment]
@husseinkorly do both certs work independently? like one at a time, do they both work?

**Comment by husseinkorly** [comment]
Only one cert can be configured in AAD, but I was assuming if we can have 2 certs and it will use the valid one. This can be useful to rotate the decryption certs without any downtime. We can add the new cert to the configuration, replace it in AAD, and remove the the old one from configuration.

**Comment by jmprieur** [comment]
@husseinkorly : provided they are both decrypt certificates, this should work.
@brentschmaltz @GeoK @mafurman : we are using `TokenValidationParameters.TokenDecryptionKeys`. Shouldn't it work?

https://github.com/AzureAD/microsoft-identity-web/blob/6b506ebb241b69dd79975b5899418af825dc1882/src/Microsoft.Identity.Web/WebApiExtensions/MicrosoftIdentityWebApiAuthenticationBuilderExtensions.cs#L222

**Comment by GeoK** [comment]
@jmprieur - Yes, this should work in IdentityModel >=6.8.0.
The library will try to decrypt a token using all decryption keys set on TokenValidationParameters.

**Comment by jennyf19** [comment]
Fixed in [1.13 release](https://github.com/AzureAD/microsoft-identity-web/releases/tag/1.13.0).

but seems to be an issue still with other components. @jmprieur

**Comment by jennyf19** [comment]
Included in [1.14 release](https://github.com/AzureAD/microsoft-identity-web/releases/tag/1.14.0)

**Comment by husseinkorly** [comment]
@jennyf19 @jmprieur 
I am not sure how this was validated, but I just got a chance to test it because we have a decryption cert expiring in few weeks. I configured the function with a second decryption cert, and it looks like the library only try the first certificate to decrypt the cert! it is not using the second cert if the first one fails.

![image](https://user-images.githubusercontent.com/3492773/148098660-b99c098e-1a24-4872-8f76-edce54705756.png)

We have other services where we pass multiple certificates, and identity model will try to all the available keys to decrypt the cert

![image](https://user-images.githubusercontent.com/3492773/148099200-333465d7-b1c3-4dfb-ba72-94f66bef0c64.png)
